Added examples for using Mudblazor with BlazorWebView / WinForms / WPF

grbd opened the pull request:
Hi,
I've recently added a couple of examples of using Mudblazor with BlazorWebView
under MudBlazor.Template.DefaultBlazor
This allows mudblazor to be used in the context of a nice looking desktop application

  * https://devblogs.microsoft.com/aspnet/asp-net-core-updates-in-net-6-preview-3/#blazorwebview-controls-for-wpf-windows-forms

The only down side currently is that it's dependent on the use of .Net 6 / VStudio 2019 preview
So I placed the examples into a separate solution.

HClausing commented on BlazorWebView.md:
````diff
@@ -0,0 +1,59 @@
+
+# Leaves the framework out
+dotnet publish -r win-x64 -c Release /p:PublishSingleFile=true --no-self-contained
+```
````
Awesome tutorial! Nice job!

grbd replied:
That's a good point, I did do that in a different project I'm working on recently so I think I'll do that here as well

HClausing replied:
> At my understanding you have made an excelent job, congratulations!
> 
> Just one suggestion: Have you ever thought in unifying the blazor content, that are repeated on WPF and WinForms into a shared project?

By the way, these shared resources could be referenced either on Server and WASM versions.

I'm intending to hitch a ride with you, and use this shared project and create a project for MAUI.

HClausing reviewed (commented):
At my understanding you have made an excelent job, congratulations!

Just one suggestion: Have you ever thought in unifying the blazor content, that are repeated on WPF and WinForms into a shared project?

grbd commented:
Ok so there should now be 4 projects under
MudBlazor.Template.DefaultBlazor\Desktop

  * DefaultBlazor.Desktop.Shared
    Common shared site code, this is similar to DefaultBlazor.Server but has some code to setup services common to all projects below
    Also has some static assests for showing a spinny loader on the winforms / wpf apps
    and uses "Microsoft.NET.Sdk.Razor" instead of "Microsoft.NET.Sdk.Web" so doesn't require cshtml pages
    
  * DefaultBlazor.Desktop.WinForms
    Example using WinForms

  * DefaultBlazor.Desktop.Wpf
    Example using Wpf

  * DefaultBlazor.Desktop.Web
    This is basically the same as using DefaultBlazor.Server but has a launch profile for hot reloading razor pages
    Also it links back to DefaultBlazor.Desktop.Shared

From what I understand BlazorWebView acts like the server side Blazor (instead of the wasm one)
although it uses IPC instead of SignalR to talk back to the host app making it faster.

I could use DefaultBlazor.Server instead of DefaultBlazor.Desktop.Shared
But there would be a couple of downsides to this

since DefaultBlazor.Server uses "Microsoft.NET.Sdk.Web" instead of "Microsoft.NET.Sdk.Razor" the WPF / WinForms examples would result in a larger build size
since they'd bring in all the web stuff associated with cshtml which isn't needed for them.
Other than that I think there's just a small code stub for setting up the services common to all and I think I added in some specific namespaces for some of the pages

HClausing reviewed (commented):
Well, I'm not as update on this lesson as you, but I'm confident that you took care on minor details for offer the best experience possible by every user that want to work with these templates.

Just one doubt: I see some files at wwwroot repeated by every target, and I wonder if it is possible to put this folder on shared project as shared asset / resource, etc.

By the way, nice job! 

Thanks!

grbd commented:
There was an erroneous test html file in the DefaultBlazor.Desktop.Web project I've just removed

For the  wwwroot/index.html files
they do appear to be the same between the winforms and the wpf apps but from what I can gather it's not possible to relocate them into DefaultBlazor.Desktop.Shared 
for info the equivalent in the DefaultBlazor.Desktop.Web example is different (that's the Pages/_Host.cshtml file)

with any static content within wwwroot under DefaultBlazor.Desktop.Shared, it doesn't get directly copied into the final build directory wwwroot (unless you publish) instead it uses some sort of special fileprovider and manifest to merge directories together.
which is why you have to use url's like _content/DefaultBlazor.Desktop.Shared/loader/spinner.svg in the index.html file.

I just tried to see if I could get away with using the following in the Blazorweb view control but it didn't seem to like it
```
HostPage = "_content/DefaultBlazor.Desktop.Shared/index.html",
```

I recently realised that MAUI doesn't support Linux, so I might try and see if I can create a Chromley demo based on
https://github.com/chromelyapps/demo-projects/tree/master/blazor/ServerAppDemo

HClausing commented:
Ok, thanks for trying that ideia. Maybe on the future the static assets will work on shared project.

There're alrerdy MAUI templates for using with VS 2019 preview (and dotnent 6 preview 5):

![image](https://user-images.githubusercontent.com/15158923/123497471-124a3b00-d604-11eb-887c-58cb63f2851b.png)

The first one if for Mobile and Mac devices, and the second is exclusive for Windows, but all the assets and classes are just links from their equivalente on first project (the blue icon on file indicates a link):

![image](https://user-images.githubusercontent.com/15158923/123497540-6bb26a00-d604-11eb-82eb-ba92fb7a8ab3.png)

I guess it will complete the template, waiting only for a Linux one.

grbd commented:
Ok so the index.html file now sits in the shared library (now renamed to DefaultBlazor.Desktop.Common)
Found a way to just label it as copy over.
although the Chromely / web frontends both use they're own _Host.cshtml file as an equivilent.

Next step is to test the chromely project under Linux (it seems to publish so far), also thinking of submitting a couple of the helper class's I wrote for that frontend to the chromely github
Then try and see if I can get Maui to work since running on Android would be cool

HClausing commented:
You arer getting many exciting results on this PR. Congratulations!

I did many tries here with MAUI + Blazor. Please let me know if you need some help with it.

On the dotnet 6 preview 5 I've discovered that Hot Reload breaks debug session on Android emulator, so we need to turn it off.

This is a screenshare of my tests:

![WhatsApp Image 2021-06-26 at 09 04 15](https://user-images.githubusercontent.com/15158923/123563266-24ee7c80-d78a-11eb-8995-b32dcd3a9bd2.jpeg)

There's a label on top using XAML language, and after the BlazorWeb View. Everything into a same XAML and Razor files / project.

grbd commented:
I've managed to get the Chromely one working on a Raspberry Pi (using arm64)
Although for some reason I had to turn off https, it seemed to get stuck generating a key for some reason otherwise

![image](https://user-images.githubusercontent.com/8965579/123656480-dee7f600-d827-11eb-956f-936fdf22b341.png)

grbd commented:
For the MAUI one I'm currently seeing if I can get the android emulator working under vmware instead to avoid turning on Hypervisor under windows (since it might hurt gaming performance)
https://linuxhint.com/vmware_android_apps_studio/

HClausing commented:
> For the MAUI one I'm currently seeing if I can get the android emulator working under vmware instead to avoid turning on Hypervisor under windows (since it might hurt gaming performance)
> https://linuxhint.com/vmware_android_apps_studio/

I'm not sure if I can help, but I've found a tutorial for just install just Android SDK and AVD without Android Studio:

https://medium.com/michael-wallace/how-to-install-android-sdk-and-setup-avd-emulator-without-android-studio-aeb55c014264

grbd commented:
Thanks for the info, I've found HAXM is actually a good option. It only works on Intel processors but it means you don't have to setup Hyper-V which can have an impact on game performance. It's also pretty easy to install

  * https://github.com/intel/haxm/releases

grbd commented:
Ok I've now added a MAUI example
  * The WinUI works (although you have to add a couple of extensions to Visual Studio to run that)
  * It also works for Android (hurray took me a while to figure out what was wrong with that)
  * There needs to be a separate index.html under wwwroot under DefaultBlazor.Desktop.Maui
    copying from the Common library doesn't work with this one, it's slightly different anyway
  * If the directory location is too deep then you'll get random errors from the Android SDK while building
  * I can deploy to an android emulator, but the debugging session seems to break after being launched for some reason. WinUI debugging works though.
  * Build feels kind of flakey, sometimes you have to clean / build / debug multiple times to get things working, but it does build / run
  * Everything is pulled from DefaultBlazor.Desktop.Common just like the other examples, making this sort of a front end to DefaultBlazor.Desktop.Common

That should be the last of the changes now
I added some notes to a Development.Maui.md file as well for things to watch out for

HClausing commented:
Thank you so much! Excelent job!

Garderoben commented:
@HClausing i had very big trouble of getting this Desktop versions to work but i was missing allot of VS extensions and what not.
Are they good to be merged?

I'm thinking of redoing the structure of this repo and have a shared resource instead its hard to make sure every thing is up today as it is now.

HClausing commented:
By the moment I had success on WPF, Winforms and MAUI/UWP for Desktop. They ran into the first try. The Chomely for Linux I need help from @grbd because didn't work here. I'm not an expert on Linux Sub-system for Windows (neither on Linux OS to be honest). Maybe some setup help on ReadMe files would be nice.

On Apple platforms I can't test because I don't have any hardware from they, even with an IOs Emulator, a phsycal hardware is mandatory for build.

On Android, There's a tip from @porkopek (thank you so much man)  that helps a lot for long path name that can be merged on `DefaultBlazor.Desktop/Development.Maui.md` file. Source: https://www.microfocus.com/documentation/filr/filr-4/filr-desktop/t47bx2ogpfz7.html

Although I think it's generally ready, it would be lovely if we were successful in testing it on Apple platforms.

grbd commented:
# MAUI

I've just added that link you mentioned for the long file paths to Development.Maui.md
That was quite handy for getting the Andoid emulator one to build

I do mention under Development.Maui.md that you need to install a couple of extensions onto Visual Studio to get the windows desktop version working

  * Project Reunion
    In the youtube clip it recommends the preview version, but I think the non preview version is now ahead with a later version number
  * Single Project MSIX Packaging Tools


# Chromely

For the Chromely one you should be able to test it under Windows as well as Linux
The finickiness it sometimes has is usually associated with it's auto download of the libcef files.

If it doesn't detect libcef in the same directory as the application it usually tries an automatic download of the libcef files
Usually it launches a browser window to say "downloading libcef files", then generates a text log on it's progress
For Linux I haven't tested it under the Windows Linux subsytem, I just copied it to an rpi and tested it from there
So the WSL may be messing it up in some unexpected way. I know that if you try and publish to a single file this also messes up the auto download of libcef

In production what you would probably do is bundle the app and libcef together anyway.
I think there are ways to also download the files manually

https://github.com/chromelyapps/Chromely/blob/master/Documents/cef_binaries_download.md

porkopek commented:
Great work, guys!
I definitively will try this. I want to migrate from Electron to MAUI. I have high expectations that everything will be smoother

Garderoben commented:
@grbd hey you still around? are you on discord would like to reach you, or this works as well.

I am completely redoing this repo structure, and making it into a real template with options, no more duplicated files etc.
would love your input on a few things, thanks.

LuohuaRain commented:
Great work, guys!
What about MAUI Preview 9？

grbd commented:
I've recently updated the templates above, I should be posting them to a separate repo later on
I'll post the link here when I've done it
Not sure about Maui, just because Maui has been a bit glitchy when I last tried it
